### PR TITLE
Impl fix bacpacmodel

### PR DESCRIPTION
**Pull request by Splaxi**

### CODE REVIEW

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+His github profile can be found here:
+        https://github.com/FH-Inway
+#>
+function Repair-D365BacpacModel {
+    [CmdletBinding()]
+    param (
+        [string] $Path,
+
+        [string] $OutputPath,
```
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1:118]: So if path is supplied, simply add `-edited` to the name, if output is empty?

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        https://github.com/batetech
+
+Florian Hopfner did a gist implementation, and did all the inital work in terms of finding the fastest way to work against the model file
+
+The original gist is: https://gist.github.com/FH-Inway/f485c720b43b72bffaca5fb6c094707e
+
+His github profile can be found here:
+        https://github.com/FH-Inway
+#>
+function Repair-D365BacpacModel {
```
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1:112]: Makes sense - will look into it

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1:1]: Yeah - tell me about it

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        $forInput = Join-Path -Path $directoryObj.FullName -ChildPath "$i.replace.input.xml"
+        $forOutput = Join-Path -Path $directoryObj.FullName -ChildPath "$i.replace.output.xml"
+
+        Copy-Item -Path $localInput -Destination $forInput -Force
+        Repair-BacpacModelReplace -Path $forInput -OutputPath $forOutput -Search $arrReplace[$i].Search -Replace $arrReplace[$i].Replace
+
+        $localInput = $forOutput
+    }
+
+    Copy-Item -Path $forOutput -Destination $OutputPath -Force
```
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1:203]: We have some cmdlets  that actually post file details - will find one to mimic

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    $localInput = Join-Path -Path $directoryObj.FullName -ChildPath "raw.replace.input.xml"
+
+    # Clone input file to the local temporary file
+    Copy-Item -Path $forOutput -Destination $localInput -Force
+
+    if (-not $KeepFiles) {
+        Get-ChildItem -Path "$($directoryObj.FullName)\*.qualifier.*.xml" | Remove-Item -Force -ErrorAction SilentlyContinue -WarningAction SilentlyContinue
+    }
+
+    for ($i = 0; $i -lt $arrReplace.Count; $i++) {
```
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1:193]: I might need some inspiration, on how to handle multiple replaces - per line....

[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1:193]: Maybe something like this:

https://devblog.pekspro.com/posts/multiple-find-and-replace-with-powershell

We have the line variable loaded entirely. So updating it, like this - would make it run....

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+As the backend of the Azure SQL infrastructure keeps evovling, the bacpac file can contain invalid instructions while we are trying to import into a local SQL Server installation on a Tier1 environment
+
+.PARAMETER Path
+Path to the bacpac model file that you want to work against
+
+.PARAMETER OutputPath
+Path to where the repaired model file should be placed
+
+.PARAMETER PathRepairSimple
+Path to the json file, that contains all the instructions to be executed in the "Simple" section
```
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1]: This has been implemented, both with a script to find the file locally - and a direct link, to the master branch, on github - for all 3 json files

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 (hunk 1 above)
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1]: Implemented

File: d365fo.tools/functions/repair-d365bacpacmodel.ps1 [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        https://github.com/FH-Inway
+#>
+function Repair-D365BacpacModel {
+    [CmdletBinding()]
+    param (
+        [string] $Path,
+
+        [string] $OutputPath,
+
+        [string] $PathRepairSimple = "$script:ModuleRoot\internal\misc\RepairBacpac.Simple.json",
```
[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1]: Great idea! I'll see what I can do...

[the remark this answers is not in the document]
Comment (Splaxi) [reply d365fo.tools/functions/repair-d365bacpacmodel.ps1]: Implemented

File: d365fo.tools/functions/repair-d365bacpacmodelfile.ps1 [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+        [switch] $KeepFiles
+    )
+
+    if ([string]::IsNullOrEmpty($OutputPath)) {
+        $OutputPath = $Path.Replace([System.IO.Path]::GetExtension($path), "-edited$([System.IO.Path]::GetExtension($path))")
+    }
+
+    $arrSimple = @()
+    if ([string]::IsNullOrEmpty($PathRepairSimple)) {
```
Comment (FH-Inway) [inline d365fo.tools/functions/repair-d365bacpacmodelfile.ps1:156]: :question:

Don't we need a `-not` here?

```suggestion
    if (-not [string]::IsNullOrEmpty($PathRepairSimple)) {
```

**Comment by Splaxi** [comment]
@FH-Inway 
- It seems that we have a good PR here. Ready for testing. Awaiting your feedback - before moving on.

**Pull request merged**